kernel/src/scan/log_replay.rs

```diff
@@ -237,21 +237,22 @@ impl LogReplayScanner {
 /// indicates whether the record batch is a log or checkpoint batch.
 pub fn scan_action_iter(
     engine: &dyn Engine,
-    action_iter: impl Iterator<Item = DeltaResult<(Box<dyn EngineData>, bool)>>,
+    action_iter: impl Iterator<Item = DeltaResult<(Box<dyn EngineData>, bool)>> + 'static,
     physical_predicate: Option<(ExpressionRef, SchemaRef)>,
-) -> impl Iterator<Item = DeltaResult<ScanData>> {
+) -> DeltaResult<impl Iterator<Item = DeltaResult<ScanData>>> {
     let mut log_scanner = LogReplayScanner::new(engine, physical_predicate);
     let add_transform = engine.get_expression_handler().get_evaluator(
         get_log_add_schema().clone(),
         get_add_transform_expr(),
         SCAN_ROW_DATATYPE.clone(),
-    );
-    action_iter
+    )?;
+
+    Ok(action_iter
```

Comment:
nit: it's very subjective, but in general I find this sort of multi-line, multi-operation `Ok` (or `Some`, etc) to be unnecessarily hard to read because of all the nesting. Factoring it out as a value is easier to read (and change later):
```rust
let actions = action_iter
      ...
    .filter(...);
Ok(actions)
```
Wrapping a single function call or struct creation isn't so bad, because conceptually only one thing is happening, e.g. this code above is not particularly difficult to read:
```rust
Ok(Arc::new(DefaultExpressionEvaluator {
    ...
}))
```
Clippy seems to take a similar stance on monadic chains -- it will almost always force newlines between chained function calls, if more than one of the functions takes args or even if the line gets very long (long before the 100-char line limit).